url542ae48576_tonyheadford_m68k_tgz-pJ8-m68k_MonitorJ8/src/m68k/cpu/instructions/ADD.java

```diff
@@ -0,0 +1,265 @@
+//       following disclaimer.
+//    o  Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//       following disclaimer in the documentation and/or other materials provided with the distribution.
+//    o  Neither the name of the M68k Project nor the names of its contributors may be used to endorse or promote
+//       products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+*/
+public class ADD implements InstructionHandler
+{
+	protected final Cpu cpu;
+
+	public ADD(Cpu cpu)
+	{
+		this.cpu = cpu;
+	}
+
+	public final void register(InstructionSet is)
+	{
+		int base;
+		Instruction i;
+
+		// destination dn
+		for(int sz = 0; sz < 3; sz++)
+		{
+			if(sz == 0)
+			{
+				// add byte (dn dest)
+				base = 0xd000;
+				i = new Instruction() {
+					public int execute(int opcode)
+					{
+						return add_byte_dn_dest(opcode);
+					}
+					public DisassembledInstruction disassemble(int address, int opcode)
+					{
+						return disassembleOp(address, opcode, Size.Byte);
+					}
+				};
+			}
+			else if(sz == 1)
+			{
+				// add word (dn dest)
+				base = 0xd040;
+				i = new Instruction() {
+					public int execute(int opcode)
+					{
+						return add_word_dn_dest(opcode);
+					}
+					public DisassembledInstruction disassemble(int address, int opcode)
+					{
+						return disassembleOp(address, opcode, Size.Word);
+					}
+				};
+			}
+			else
+			{
+				// add long (dn dest)
+				base = 0xd080;
+				i = new Instruction() {
+					public int execute(int opcode)
+					{
+						return add_long_dn_dest(opcode);
+					}
+					public DisassembledInstruction disassemble(int address, int opcode)
+					{
+						return disassembleOp(address, opcode, Size.Long);
+					}
+				};
+			}
+			for(int reg = 0; reg < 8; reg++)
+			{
+				for(int ea_mode = 0; ea_mode < 8; ea_mode++)
+				{
+					for(int ea_reg = 0; ea_reg < 8; ea_reg++)
+					{
+						if(ea_mode == 7 && ea_reg > 4)
+							break;
+						is.addInstruction(base + (reg << 9) + (ea_mode << 3) + ea_reg, i);
+					}
+				}
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 91:102, 153:164.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.